File: src/main/java/com/tabnine/inline/render/TabnineInlay.java
package com.tabnine.inline.render;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.editor.Editor;
import com.tabnine.capabilities.CapabilitiesService;
import com.tabnine.capabilities.Capability;
import com.tabnine.inline.render.experimental.ExperimentalTabnineInlay;
import com.tabnine.inline.render.preserved.DefaultTabnineInlay;
import com.tabnine.prediction.TabNineCompletion;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

public interface TabnineInlay {
    Integer getOffset();
    Rectangle getBounds();
    boolean isEmpty();
    void register(Disposable parent);
    void clear();
    void render(Editor editor, @NotNull String suffix, TabNineCompletion completion, int offset);

    static TabnineInlay create() {
        if (CapabilitiesService.getInstance().isCapabilityEnabled(Capability.ALPHA)) {
            return new ExperimentalTabnineInlay();
        }

        return new DefaultTabnineInlay();
    }
}

File: src/main/java/com/tabnine/inline/render/TabnineInlay.kt
package com.tabnine.inline.render

import com.intellij.openapi.Disposable
import com.intellij.openapi.editor.Editor
import com.tabnine.prediction.TabNineCompletion
import java.awt.Rectangle

interface TabnineInlay {
    val offset: Int?
    val isEmpty: Boolean

    fun getBounds(): Rectangle?
    fun register(parent: Disposable)
    fun clear()
    fun render(editor: Editor, completion: TabNineCompletion, offset: Int)

    companion object {
        @JvmStatic
        fun create(): TabnineInlay {
            return DefaultTabnineInlay()
        }
    }
}
